pkg/scheduler/core/generic_scheduler_test.go

```diff
@@ -2259,3 +2259,43 @@ func assignDefaultStartTime(pods []*v1.Pod) {
 		}
 	}
 }
+
+func TestFairEvaluationForNodes(t *testing.T) {
+	defer algorithmpredicates.SetPredicatesOrderingDuringTest(order)()
+	predicates := map[string]algorithmpredicates.FitPredicate{"true": truePredicate}
+	numAllNodes := 500
+	nodeNames := make([]string, 0, numAllNodes)
+	for i := 0; i < numAllNodes; i++ {
+		nodeNames = append(nodeNames, strconv.Itoa(i))
+	}
+	nodes := makeNodeList(nodeNames)
+	g := makeScheduler(predicates, nodes)
+	g.percentageOfNodesToScore = 30
+	nodesToFind := int(g.numFeasibleNodesToFind(int32(numAllNodes)))
+
+	// Iterating over all nodes more than twice
+	for i := 0; i < 2*(numAllNodes/nodesToFind+1); i++ {
+		prev := g.lastProcessedNodeIndex
+		nodesThatFit, _, _, err := g.findNodesThatFit(context.Background(), framework.NewCycleState(), &v1.Pod{})
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+		if len(nodesThatFit) != nodesToFind {
+			t.Errorf("got %d nodes filtered, want %d", len(nodesThatFit), nodesToFind)
+		}
+		if g.lastProcessedNodeIndex != (i+1)*nodesToFind%numAllNodes {
+			t.Errorf("got %d lastProcessedNodeIndex, want %d", g.lastProcessedNodeIndex, (i+1)*nodesToFind%numAllNodes)
+		}
+		for _, node := range nodesThatFit {
+			idx, err := strconv.Atoi(node.Name)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			// We can only check the lower bound, the upper bound is non-deterministic.
+			if prev < g.lastProcessedNodeIndex && idx < prev {
```

Comment:
This conditional doesn't really make sense, because we are indexing in circle. I'm ok with just removing this check.
Otherwise, the check should look like this:
`prev < idx <= g.lastProcessedNodeIndex`, but in a circle.

Reply:
> This conditional doesn't really make sense, because we are indexing in circle. I'm ok with just removing this check.

What we want to test is whether the evaluation range actually slides.  So this check is useful in the case someone will happen to change to not use `lastProcessNodeIndex` to determine the range but `lastProcessNodeIndex` is still updating.

Of course, the range is in circle.  But as in the comment, we cannot ensure the upper bound, so current check is compromised.

Reply:
We can't do it in general or we can't do it because it's in a circle? I think the later (which is just a matter of adjusting the conditionals)

Reply:
We can't ensure the upper bound, because we use `workqueue. ParallelizeUntil`. 
So the check is `prev <= idx` when index is increasing constantly(`prev < g.lastProcessedNodeIndex`). If the range includes index 0, we can't check anything.
And I think the check is not `prev < idx` but  `prev <= idx`. The field name `lastPrcessedNodeIndex` is not correct. It shoud be `startPointNodeIndex` or so.

Reply:
Right, given the parallelism, the indexes might not be in the range.
In that sense, I prefer that we just give up on this check altogether (lines 2289 to 2299). This is just a non-deterministic heuristic to make scheduling faster across cycles.

And I'm ok with changing the name of the variable.

cc @ahg-g

Reply:
I think this  (lines 2289 to 2299) is a deterministic check, but it's ok to remove it. And I changed the variable name.

Reply:
it was a deterministic check of a non-deterministic heuristic. But it was also incomplete due to the circular nature of the algorithm.

Thanks for changing the variable name.